ExpressionUtils/Evaluating/CachedExpressionCompiler.cs

```diff
@@ -82,7 +83,29 @@ private static object getDefaultValue(Type t) {
 		/// Use for testing only.
 		/// </remarks>
 		internal bool IsCached(LambdaExpression lambda) {
-			return delegates.ContainsKey(lambda);
+			return delegates.ContainsKey((lambda.Body, lambda.Parameters));
+		}
+
+		/// <summary>
+		/// Previously as a key for <see cref="delegates"/> we were using <see cref="LambdaExpression"/> objects and <see cref="ExpressionComparing.StructuralComparer"/> as a comparer.
+		/// But that required us to make calls to <see cref="Expression.Lambda(Expression, ParameterExpression[])"/> which contains global lock inside, and that started to become a problem.
+		/// So instead we now use <see cref="(Expression , IReadOnlyCollection{ParameterExpression})"/> tuple as a key and pass body and parameters separately, to reduce number of calls to <see cref="Expression.Lambda(Expression, ParameterExpression[])"/>.
+		/// </summary>
+		private class LambdaExpressionComparer : IEqualityComparer<(Expression body, IReadOnlyCollection<ParameterExpression> parameters)> {
+			private IEqualityComparer<Expression> expressionComparer = new ExpressionComparing.StructuralComparer(ignoreConstantsValues: true);
+
+			public bool Equals((Expression body, IReadOnlyCollection<ParameterExpression> parameters) x, (Expression body, IReadOnlyCollection<ParameterExpression> parameters) y) {
+				return x.body.Type == y.body.Type
+					&& x.parameters.SequenceEqualOrBothNull(y.parameters, expressionComparer.Equals)
+					&& expressionComparer.Equals(x.body, y.body);
+			}
+
+			public int GetHashCode((Expression body, IReadOnlyCollection<ParameterExpression> parameters) obj) {
+				var hash = Hashing.FnvOffset;
+				Hashing.Hash(ref hash, obj.body.Type.GetHashCode());
+				Hashing.Hash(ref hash, expressionComparer.GetHashCode(obj.body));
+				return hash;

```

Comment:
Is it intended that you don't include `parameters` in the hash?

Reply:
Yes, i stick to the same logic that was here before trying to keep any logic changes to minimum. 

So previously `StructuralComparer.GetHashCode()` was used for hash. It just called `HashCodeVisitor` to do the actual work. And when `HashCodeVisitor` gets expression that is `LambdaExpression` it just adds hash of return type and then visits body. So in our case we add to hash `body.Type` which should be the same as lambda return type and the call visitor on body. As you can see parameters were not used before. So i basically to the same thing that was done before.
<img width="618" alt="image" src="https://github.com/user-attachments/assets/428ef8c4-b88a-4e1f-8211-4697067cc938">

Reply:
Okay, makes sense